comps/llms/text-generation/vllm_ray/vllm_openai.py

```diff
@@ -0,0 +1,183 @@
+
+            initialize_distributed_hpu()
+            torch.zeros(1).to("hpu")
+            device = "HPU"
+        except Exception:
+            device = "GPU"
+    parsed_args = parse_vllm_args(cli_args)
+    engine_args = AsyncEngineArgs.from_cli_args(parsed_args)
+    engine_args.worker_use_ray = True
+
+    tp = engine_args.tensor_parallel_size
+    logger.info(f"Tensor parallelism = {tp}")
+    pg_resources = []
+    pg_resources.append({"CPU": 1})  # for the deployment replica
+    for i in range(tp):
+        pg_resources.append({"CPU": 1, device: 1})  # for the vLLM actors
+
+    # We use the "STRICT_PACK" strategy below to ensure all vLLM actors are placed on
+    # the same Ray node.
+    return VLLMDeployment.options(placement_group_bundles=pg_resources, placement_group_strategy="STRICT_PACK").bind(
+        engine_args,
+        parsed_args.response_role,
+        parsed_args.lora_modules,
+        parsed_args.chat_template,
+    )
+
+
+# __serve_example_end__
+
+
+def main(argv=None):
+    import argparse
+
+    parser = argparse.ArgumentParser(description="Serve vLLM models with Ray Serve.", add_help=True)
+    parser.add_argument("--port_number", default=8008, type=int, help="Port number to serve on.")
+    parser.add_argument(
+        "--model_id_or_path", default="NousResearch/Meta-Llama-3-8B-Instruct", type=str, help="Model id or path."
+    )
+    parser.add_argument(
+        "--chat_processor", default="ChatModelNoFormat", type=str, help="Chat processor for aligning the prompts."
+    )
+    parser.add_argument("--tensor_parallel_size", default=1, type=int, help="parallel nodes number for 'hpu' mode.")
+
+    if len(sys.argv) == 1:
+        parser.print_help(sys.stderr)
+        sys.exit(1)
+
+    args = parser.parse_args(argv)
+
+    serve.run(
+        build_app({"model": args.model_id_or_path, "tensor-parallel-size": args.tensor_parallel_size, "device": "HPU"})
+    )
+    # __query_example_begin__
+    # from openai import OpenAI
+
+    # # Note: Ray Serve doesn't support all OpenAI client arguments and may ignore some.
+    # client = OpenAI(
+    #     # Replace the URL if deploying your app remotely
+    #     # (e.g., on Anyscale or KubeRay).
+    #     base_url="http://localhost:8000/v1",
+    #     api_key="NOT A REAL KEY",
+    # )
+    # chat_completion = client.chat.completions.create(
+    #     model="NousResearch/Meta-Llama-3-8B-Instruct",
+    #     messages=[
+    #         {"role": "system", "content": "You are a helpful assistant."},
+    #         {
+    #             "role": "user",
+    #             "content": "What are some highly rated restaurants in San Francisco?'",
+    #         },
+    #     ],
+    #     temperature=0.01,
+    #     stream=True,
+    # )
+
+    # for chat in chat_completion:
+    #     if chat.choices[0].delta.content is not None:
+    #         print(chat.choices[0].delta.content, end="")
+    # __query_example_end__
```

Comment:
It looks like we'll use microservice to send requests to server, please remove these if they are not needed here.

Reply:
removed.